### PR TITLE
Fixed Equals method for transformers

### PR DESCRIPTION
**Pull request by bahusoid**
Fixed possible equality issue:
```C#
public class MyTransformer: DistinctRootEntityResultTransformer {}
var myTransformer = new MyTransformer();
var distinctTransformer = new DistinctRootEntityResultTransformer();
myTransformer.Equals(distinctTransformer); // true. Should be false
distinctTransformer.Equals(myTransformer); // false
```
Also reuse instances when possible

### CODE REVIEW

File: src/NHibernate.Test/NHSpecificTest/NH3957/ResultTransformerEqualityFixture.cs [hunk 1]
```diff
@@ -8,55 +7,25 @@ namespace NHibernate.Test.NHSpecificTest.NH3957
...
-		{
-			// Restore those types hashcode. (Avoid impacting perf of other tests, avoid second level query cache
-			// issues if it was holding cached entries (but would mean some tests have not cleaned up properly).)
-			foreach(var backup in _hasherBackup)
-			{
-				TweakHashcode(backup.Key, backup.Value);
-			}
-		}
-
+		public class CustomAliasToEntityMapResultTransformer : AliasToEntityMapResultTransformer { }
```
Comment (bahusoid) [inline src/NHibernate.Test/NHSpecificTest/NH3957/ResultTransformerEqualityFixture.cs]: This custom class provides also hashcode collision with current implementaion. Not sure why TweakHashcode was required

Comment (fredericDelaporte) [reply src/NHibernate.Test/NHSpecificTest/NH3957/ResultTransformerEqualityFixture.cs]: See #564, it is required to check that two different result transformers will not be considered equal if they happen to have the same hashcode, which was NH-3957 bug. The collision for different instances of the same transformer is intentional and mandatory since they are considered equal.

Removing the tweak causes the `Inequality` suffixed tests to no more test the NH-3957 bug. You should change them to compare with the corresponding `Custom` descendant you have added. That is still a step away from what was NH-3957, but I do not see any other solution with the `Instance` optimization. NH-3957 was a random hashcode collision between different unrelated transformers causing them to wrongly consider themselves equals.

Comment (bahusoid) [reply src/NHibernate.Test/NHSpecificTest/NH3957/ResultTransformerEqualityFixture.cs]: >it is required to check that two different result transformers will not be considered equal if they happen to have the same hashcode,.. Removing the tweak causes the Inequality suffixed tests to no more test the NH-3957 bug

But I'm already checking `Inequality`:
[`Assert.False(custom.Equals(transf1));`](https://github.com/nhibernate/nhibernate-core/pull/2000/files#diff-012db71c0b0b1c67464a20c007d76d40R53)
[`Assert.False(transf1.Equals(custom));`](https://github.com/nhibernate/nhibernate-core/pull/2000/files#diff-012db71c0b0b1c67464a20c007d76d40R54)

 I've just put those checks in the wrong tests.
>That is still a step away from what was NH-3957

I still see no difference between inequality check for custom implemented classes with collisions to existing ones ( `custom` and `transf1`) and TweakHashcode hack for existing classes. IMHO they still fully cover the case:
>that two different result transformers will not be considered equal if they happen to have the same hashcode

File: src/NHibernate/Linq/IntermediateHqlTree.cs [hunk 2]
```diff
@@ -110,7 +110,7 @@ public void AddDistinctRootOperator()
 			if (!_hasDistinctRootOperator)
 			{
 				Expression<Func<IEnumerable<object>, IList>> x =
-					l => new DistinctRootEntityResultTransformer().TransformList(l.ToList());
+					l => DistinctRootEntityResultTransformer.Instance.TransformList(l.ToList());
```
Comment (bahusoid) [inline src/NHibernate/Linq/IntermediateHqlTree.cs]: I also want to get rid of this ToList conversion. Will undo this change and make it in my other PR
